remote_attestation/rust/src/tests/message.rs

```diff
@@ -0,0 +1,222 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    crypto::{
+        KEY_AGREEMENT_ALGORITHM_KEY_LENGTH, NONCE_LENGTH, SIGNATURE_LENGTH,
+        SIGNING_ALGORITHM_KEY_LENGTH,
+    },
+    message::{
+        deserialize_message, ClientHello, ClientIdentity, Deserializable, EncryptedData,
+        MessageWrapper, Serializable, ServerIdentity, MAXIMUM_MESSAGE_SIZE,
+        REPLAY_PROTECTION_ARRAY_LENGTH,
+    },
+};
+use anyhow::{anyhow, Context};
+use assert_matches::assert_matches;
+use quickcheck::{quickcheck, TestResult};
+
+const INVALID_MESSAGE_HEADER: u8 = 5;
+
+/// Creates a zero initialized array.
+fn default_array<T, const L: usize>() -> [T; L]
+where
+    T: std::marker::Copy + std::default::Default,
+{
+    [Default::default(); L]
+}
+
+/// Converts slices to arrays (expands with zeroes).
+fn to_array<T, const L: usize>(input: &[T]) -> anyhow::Result<[T; L]>
+where
+    T: std::marker::Copy + std::default::Default,
+{
+    if input.len() <= L {
+        let mut result: [T; L] = default_array();
+        result[..input.len()].copy_from_slice(&input[..input.len()]);
+        Ok(result)
+    } else {
+        Err(anyhow!(
+            "Maximum input length exceeded, maximum {}, found {}",
+            L,
+            input.len()
+        ))
+    }
+}
+
+fn test_serialize_template<M: Serializable + Deserializable + PartialEq>(
+    message: &M,
+) -> anyhow::Result<bool> {
+    let serialized_message = message.serialize().context("Couldn't serialize message")?;
```

Comment:
I try to put assertions only to top-level functions marked as `#[test]`, so that it's easier to read errors in the test output (i.e. is nested functions are involved, `anyhow` will show them).

`TestResult` is necessary to allow discarding test values that don't sutisfy some preconditions.